lib/vrl/stdlib/src/find_table_row.rs

```diff
@@ -0,0 +1,191 @@
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "condition",
+                kind: kind::OBJECT,
+                required: true,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[]
+    }
+
+    fn compile(&self, state: &state::Compiler, mut arguments: ArgumentList) -> Compiled {
+        let tables = state
+            .get_enrichment_tables()
+            .as_ref()
+            .map(|tables| {
+                tables
+                    .table_ids()
+                    .into_iter()
+                    .map(Value::from)
+                    .collect::<Vec<_>>()
+            })
+            .unwrap_or_else(Vec::new);
+
+        let table = arguments.required_enum("table", &tables)?.to_string();
+        let condition = arguments.required_object("condition")?;
+
+        Ok(Box::new(FindTableRowFn { table, condition }))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct FindTableRowFn {
+    table: String,
+    condition: BTreeMap<String, expression::Expr>,
+}
+
+impl Expression for FindTableRowFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let condition = self
+            .condition
+            .iter()
+            .map(|(key, value)| {
+                Ok(Condition::Equals {
+                    field: key.to_string(),
+                    value: value.resolve(ctx)?.try_bytes_utf8_lossy()?.to_string(),
+                })
+            })
+            .collect::<Result<Vec<Condition>>>()?;
+
+        let tables = ctx
+            .get_enrichment_tables()
+            .as_ref()
+            .ok_or("enrichment tables not loaded")?;
+
+        match tables.find_table_row(&self.table, condition)? {
+            None => Err("data not found".into()),
+            Some(data) => Ok(Value::Object(data)),
+        }
+    }
+
+    fn update_state(
+        &self,
+        state: &mut state::Compiler,
+    ) -> std::result::Result<(), ExpressionError> {
+        match state.get_enrichment_tables_mut() {
+            Some(ref mut table) => {
+                let fields = self
+                    .condition
+                    .iter()
+                    .map(|(field, _)| field.as_ref())
+                    .collect::<Vec<_>>();
+                table.add_index(&self.table, fields)?;
+                Ok(())
+            }
+            // We shouldn't reach this point since the type checker will ensure the table exists before this function is called.
+            None => Err("enrichment tables aren't loaded".into()),
+        }
+    }
+
+    fn type_def(&self, _: &state::Compiler) -> TypeDef {
+        TypeDef::new()
+            .fallible()
+            .add_object::<(), Kind>(map! { (): Kind::Bytes })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use shared::{btreemap, TimeZone};
+    use vrl::{Condition, EnrichmentTableSearch, EnrichmentTableSetup};
```

Comment:
Are all three of these types related to enrichment tables?

If so, I'd prefer it if we namespace them accordingly:

```suggestion
    use vrl::enrichment::{Condition, TableSearch, TableSetup};
```

Reply:
I think `vrl::Condition` is generic - the other two could be namespaced.

Reply:
`Condition` here does apply only to enrichment tables, there are other `Condition`s about such as the Vector ones which are different things. Definitely suggests namespacing it is a good idea to avoid confusion.